src/libexpr/parser-state.hh

```diff
@@ -20,25 +20,58 @@ struct StringToken
 
 struct ParserLocation
 {
-    int first_line, first_column;
-    int last_line, last_column;
+    int beginOffset;
+    int endOffset;
 
     // backup to recover from yyless(0)
-    int stashed_first_column, stashed_last_column;
+    int stashedBeginOffset, stashedEndOffset;
 
     void stash() {
-        stashed_first_column = first_column;
-        stashed_last_column = last_column;
+        stashedBeginOffset = beginOffset;
+        stashedEndOffset = endOffset;
     }
 
     void unstash() {
-        first_column = stashed_first_column;
-        last_column = stashed_last_column;
+        beginOffset = stashedBeginOffset;
+        endOffset = stashedEndOffset;
     }
+
+    /** Latest doc comment position, or 0. */
+    int doc_comment_first_column, doc_comment_last_column;
+};
+
+struct LexerState
+{
+    /**
+     * Tracks the distance to the last doc comment, in terms of lexer tokens.
+     *
+     * The lexer sets this to 0 when reading a doc comment, and increments it
+     * for every matched rule; see `lexer-helpers.cc`.
+     * Whitespace and comment rules decrement the distance, so that they result
+     * in a net 0 change in distance.
+     */
+    int docCommentDistance = 9999;
+    // We're not using this whole struct actually
+    ParserLocation lastDocCommentLoc;
+
+    /**
+     * @brief Parser location to DocComment
+     *
+     * The shared_ptr allows us to have a single instance for each comment,
+     * so that it can be updated with things like a name from a binding during
+     * parsing.
+     */
+    std::map<PosIdx, DocComment> & positionToDocComment;
```

Comment:
Maybe have the parser return this instead of passing a pointer.

Reply:
This needs to be accessed during parsing, which really limits our options of how we can do this.

Also it wouldn't be wrong to use this with a "global" map instead. It's an `EvalState` choice not to do this, for the benefit of trivial thread safety of this map during (potentially concurrent) parsing.